File: system-test/java/src/main/java/com/optivem/shop/testkit/driver/adapter/shared/client/playwright/PageClient.java
package com.optivem.shop.testkit.driver.adapter.shared.client.playwright;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.WaitForSelectorState;

import java.util.List;

public class PageClient {
    // Increased default timeout for parallel test execution
    private static final int DEFAULT_TIMEOUT_SECONDS = 30;
    private static final int DEFAULT_TIMEOUT_MILLISECONDS = DEFAULT_TIMEOUT_SECONDS * 1000;
    private final Page page;
    private final int timeoutMilliseconds;

    public PageClient(Page page) {
        this(page, DEFAULT_TIMEOUT_MILLISECONDS);
    }

    public PageClient(Page page, int timeoutMilliseconds) {
        this.page = page;
        this.timeoutMilliseconds = timeoutMilliseconds;
    }

    public void fill(String selector, String value) {
        var locator = getLocator(selector);
        var processedValue = value == null ? "" : value;
        locator.fill(processedValue);
    }

    public void click(String selector) {
        var locator = getLocator(selector);
        locator.click();
    }

    public String readTextContent(String selector) {
        var locator = getLocator(selector);
        return locator.textContent();
    }

    public String readAttribute(String selector, String attribute) {
        var locator = getLocator(selector);
        return locator.getAttribute(attribute);
    }

    public List<String> readAllTextContents(String selector) {
        var locator = page.locator(selector);
        // Wait for at least one element to be visible
        // allTextContents() doesn't trigger strict mode - it's designed for multiple elements
        locator.first().waitFor(getDefaultWaitForOptions());
        return locator.allTextContents();
    }

    public List<String> readAllTextContentsNoWait(String selector) {
        var locator = page.locator(selector);
        return locator.allTextContents();
    }

    public void waitForVisible(String selector) {
        var locator = page.locator(selector);
        locator.waitFor(getDefaultWaitForOptions());
    }

    public boolean isVisible(String selector) {
        try {
            var locator = getLocator(selector);
            return locator.count() > 0;
        } catch (Exception e) {
            return false;
        }
    }

    public void removeElements(String selector) {
        page.evaluate("selector => document.querySelectorAll(selector).forEach(el => el.remove())", selector);
    }

    public boolean isHidden(String selector) {
        var locator = page.locator(selector);
        return locator.count() == 0;
    }

    private Locator getLocator(String selector, Locator.WaitForOptions waitForOptions) {
        var locator = page.locator(selector);
        locator.waitFor(waitForOptions);

        if (locator.count() == 0) {
            throw new IllegalStateException("No elements found for selector: " + selector);
        }

        return locator;
    }

    private Locator getLocator(String selector) {
        var waitForOptions = getDefaultWaitForOptions();
        return getLocator(selector, waitForOptions);
    }

    private Locator.WaitForOptions getDefaultWaitForOptions() {
        return new Locator.WaitForOptions()
                .setState(WaitForSelectorState.VISIBLE)
                .setTimeout(timeoutMilliseconds);
    }
}

File: system-test/typescript/src/testkit/driver/adapter/shared/playwright/PageClient.ts
import type { Page } from '@playwright/test';

const DEFAULT_TIMEOUT_MS = 5000;

export class PageClient {
  constructor(
    private readonly page: Page,
    private readonly timeoutMs: number = DEFAULT_TIMEOUT_MS
  ) {}

  async goto(url: string): Promise<void> {
    await this.page.goto(url);
  }

  async fill(selector: string, value: string): Promise<void> {
    await this.page.fill(selector, value, { timeout: this.timeoutMs });
  }

  async click(selector: string): Promise<void> {
    await this.page.click(selector, { timeout: this.timeoutMs });
  }

  async readTextContent(selector: string): Promise<string> {
    const text = await this.page.textContent(selector, { timeout: this.timeoutMs });
    return text ?? '';
  }

  async readAttribute(selector: string, attribute: string): Promise<string | null> {
    return this.page.getAttribute(selector, attribute, { timeout: this.timeoutMs });
  }

  async readAllTextContents(selector: string): Promise<string[]> {
    await this.page.waitForSelector(selector, { timeout: this.timeoutMs });
    return this.page.locator(selector).allTextContents();
  }

  async readAllTextContentsNoWait(selector: string): Promise<string[]> {
    return this.page.locator(selector).allTextContents();
  }

  async waitForVisible(selector: string): Promise<void> {
    await this.page.waitForSelector(selector, { state: 'visible', timeout: this.timeoutMs });
  }

  async isVisible(selector: string): Promise<boolean> {
    return this.page.locator(selector).isVisible();
  }

  async isHidden(selector: string): Promise<boolean> {
    return this.page.locator(selector).isHidden();
  }

  getPage(): Page {
    return this.page;
  }
}
